src/components/MyData/SleepChart/useSleepChartData.ts

```diff
@@ -0,0 +1,88 @@
+import { useEffect, useState } from 'react';
+import { useFhirClient } from '../../../hooks';
+import { ScaleTime, scaleTime } from 'd3-scale';
+import { differenceInDays, min, max, startOfDay, endOfDay } from 'date-fns';
+import { useCommonChartProps } from '../useCommonChartProps';
+import { Observation } from 'fhir/r3';
+import compact from 'lodash/compact';
+
+type Props = {
+  dateRange: [Date, Date];
+};
+
+export type SleepChartData = ReturnType<typeof useSleepChartData>;
+
+export const useSleepChartData = (props: Props) => {
+  const { dateRange } = props;
+  const { useSearchResourcesQuery } = useFhirClient();
+  const common = useCommonChartProps();
+  const [chartData, setChartData] = useState<{
+    sleepData: Observation[];
+    xDomain: ScaleTime<number, number>;
+    dateRange: [Date, Date];
+  }>({ sleepData: [], xDomain: scaleTime(), dateRange });
+  const [isLoading, setIsLoading] = useState(true);
+
+  const days = Math.abs(differenceInDays(...dateRange));
+
+  const { data, isFetching, isFetched } = useSearchResourcesQuery({
+    resourceType: 'Observation',
+    coding: [
+      {
+        system: 'http://snomed.info/sct',
+        code: '258158006',
+      },
+    ],
+    dateRange: [startOfDay(dateRange[0]), endOfDay(dateRange[1])],
+    pageSize: Math.max(50, days),
+  });
+
+  useEffect(() => {
+    if (!isFetching && isFetched) {
+      const newSleepData =
+        data?.entry
+          ?.map((e) => e.resource)
+          .filter((v): v is Observation => !!v) ?? [];
+
+      const { start, end } = newSleepData.reduce(
+        (domain, observation) => ({
+          start: min(
+            compact([
+              domain.start,
+              observation.valuePeriod?.start &&
+                new Date(observation.valuePeriod.start),
+              ...(observation.component?.map(
+                (c) => c.valuePeriod?.start && new Date(c.valuePeriod?.start),
+              ) ?? []),
+            ]),
+          ),
+          end: max(
+            compact([
+              domain.end,
+              observation.valuePeriod?.end &&
+                new Date(observation.valuePeriod.end),
+              ...(observation.component?.map(
+                (c) => c.valuePeriod?.end && new Date(c.valuePeriod?.end),
+              ) ?? []),
+            ]),
+          ),
+        }),
+        {} as { start?: Date; end?: Date },
+      );
+
+      const newXDomain = scaleTime()
+        .range([0, common.plotAreaWidth])
+        .domain([new Date(start ?? 0), new Date(end ?? 0)]);
+
+      setChartData({ sleepData: newSleepData, xDomain: newXDomain, dateRange });
```

Comment:
Like we've talked about, whether in this hook or elsewhere, I foresee the need to filter `newSleepData` down to handle multiple sleep records per evening (e.g. one from Oura, one from Garmin -> both synced in from Apple Health).  Unless that logic is already in place elsewhere, just keep it in mind (as perhaps a separate task to eventually get to).  Excited to try this out with my own data, which would serve as an example of that scenario.

Reply:
I kind of forgot about that discussion. I think it will likely have some display issues with the current solution. I look into it and follow up with a fix.